security/s2a/internal/crypter/aesgcm_test.go

```diff
@@ -0,0 +1,513 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package crypter
+
+import (
+	"bytes"
+	"fmt"
+	. "google.golang.org/grpc/security/s2a/internal/crypter/testutil"
+	"testing"
+)
+
+// getGCMCryptoPair outputs a sender/receiver pair on AES-GCM.
+func getGCMCryptoPair(key []byte, t *testing.T) (S2AAeadCrypter, S2AAeadCrypter) {
+	sender, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(ClientSide, key) = %v", err)
+	}
+	receiver, err := NewAESGCM(key)
+	if err != nil {
+		t.Fatalf("NewAESGCM(ServerSide, key) = %v", err)
+	}
+	return sender, receiver
+}
+
+func isFailure(result string, err error, got, expected []byte) bool {
+	return (result == ValidResult && (err != nil || !bytes.Equal(got, expected))) ||
+		(result == InvalidResult && bytes.Equal(got, expected))
+}
+
+func wycheProofTestVectorFilter(testGroup TestGroup) bool {
+	return testGroup.IvSize != 96 ||
+		(testGroup.KeySize != 128 && testGroup.KeySize != 256) ||
+		testGroup.TagSize != 128
+}
+
+func testGCMEncryptionDecryption(sender S2AAeadCrypter, receiver S2AAeadCrypter, test *CryptoTestVector, t *testing.T) {
+	// Ciphertext is: encrypted text + tag.
+	var ciphertext []byte
+	ciphertext = append(ciphertext, test.Ciphertext...)
+	ciphertext = append(ciphertext, test.Tag...)
+
+	// Decrypt.
+	got, err := receiver.Decrypt(nil, ciphertext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, test.Plaintext) {
+		t.Errorf("key=%v\ntag=%v\nciphertext=%v\nDecrypt = %v, %v\nwant: %v",
+			test.Key, test.Tag, ciphertext, got, err, test.Plaintext)
+	}
+
+	// Encrypt.
+	var dst []byte
+	if test.AllocateDst {
+		dst = make([]byte, len(test.Plaintext)+sender.TagSize())
+	}
+	got, err = sender.Encrypt(dst[:0], test.Plaintext, test.Nonce, test.Aad)
+	if isFailure(test.Result, err, got, ciphertext) {
+		t.Errorf("key=%v\nplaintext=%v\nEncrypt = %v, %v\nwant: %v",
+			test.Key, test.Plaintext, got, err, ciphertext)
+	}
+}
+
+func testGCMEncryptRoundtrip(sender S2AAeadCrypter, receiver S2AAeadCrypter, t *testing.T) {
+	// Construct a dummy nonce.
+	nonce := make([]byte, NonceSize)
+
+	// Encrypt.
+	const plaintext = "This is plaintext."
+	var err error
+	buf := []byte(plaintext)
+	buf, err = sender.Encrypt(buf[:0], buf, nonce, nil)
```

Comment:
I don't think we want to do this as this will be testing a different part of the code than what we want. Here, are reusing the buffer as both the input and the output (hence, why buf represents both ciphertext as well as plaintext throughout it's lifetime). I think this is intended to test the case where the input fully overlaps with the output. WDYT?

Reply:
This is fine. Thanks for the clarification. Can you please document this in a comment so we don't forget about it in the future?

Reply:
Done